docs/docs/tutorials/jwt_authentication.md

```diff
@@ -0,0 +1,515 @@
+
+```bash
+curl -X POST \
+     -H "Content-Type: application/json" \
+     -d '{"username": "john", "password": "123"}'  \
+     http://localhost:8080/sign_in
+
+# {"token":"john"}
+```
+
+## Requiring authentication to access routes
+
+Now that we have the means to get an authentication token, we can now protect routes to require a token
+to be provided in
+order to access them.
+
+To start lets create the following route:
+
+```dart
+// routes/tasks/index.dart
+import 'dart:io';
+
+import 'package:dart_frog/dart_frog.dart';
+
+Future<Response> onRequest(RequestContext context) {
+  return switch (context.request.method) {
+    HttpMethod.post => _onPost(context),
+    _ => Future.value(Response(statusCode: HttpStatus.methodNotAllowed)),
+  };
+}
+
+Future<Response> _onPost(RequestContext context) async {
+  final task = await context.request.body();
+  return Response.json(
+    body: {
+      'recorded_task': task,
+    },
+  );
+}
+```
+
+This route doesn't do much for now, it just reads the request and answers with a body with the
+received task name. Also it isn't yet protected.
+
+To protect our route, we will use [`dart_frog_auth`](https://pub.dev/packages/dart_frog_auth), a
+package provided by Dart Frog that makes it easier to implement token based authentications.
+
+So lets start by adding it to the project:
+
+```bash
+dart pub add dart_frog_auth
+```
+
+First, we need to create a method in our authenticator class which will be responsible to validate
+an authentication token, since right now, our authentication token is just the user's username,
+we can add the following snippet to the `Authenticator` class:
+
+```dart
+  User? verifyToken(String username) {
+    return _users[username];
+  }
+```
+
+If the token is valid, the user will be returned, otherwise, the method will return `null`.
+
+Next, lets create the following middleware under `routes/tasks`:
+
+```dart
+import 'package:authenticated_app/authenticator.dart';
+import 'package:authenticated_app/user.dart';
+import 'package:dart_frog/dart_frog.dart';
+import 'package:dart_frog_auth/dart_frog_auth.dart';
+
+Handler middleware(Handler handler) {
+  return handler.use(
+    bearerAuthentication<User>(
+      authenticator: (context, token) async {
+        final authenticator = context.read<Authenticator>();
+        return authenticator.verifyToken(token);
+      },
+    ),
+  );
+}
+```
+
+What this middleware does, is to add a Bearer token authentication checking to all requests incoming
+into the routes of that namespace.
+
+The usage of `bearerAuthentication` middleware, which is provided by `dart_frog_auth` is quite simple.
+We simply need to inform the a function to the `bearerAuthentication` attribute, this function receives
+the current `RequestContext` and the token that was passed in the request.
+
+If the token is valid and belongs to a user, the function must return that user. Otherwise,
+it should return null.
```

Comment:
Adjusted to avoid using "simply" twice:

```suggestion
We need to provide a function that returns a user if the token is valid and belongs to a user; or null otherwise.
```